### PR TITLE
Fix import logic with --file for dhall-to-{json,yaml}

### LINKED ISSUE #1183: `--file` resolves paths differently between `dhall` and `dhall-to-yaml`
Possibly related to https://github.com/dhall-lang/dhall-haskell/pull/1004

I have a directory structure like so:
```
$ tree foo
foo
├── one.dhall
└── two.dhall
```
With the two files:
```
$ cat foo/one.dhall
{ foo = "bar" }

$ cat foo/two.dhall
let foo = ./one.dhall

in
{ baz = "buz"
, foo = foo.foo
}
```

When I use `dhall --file` to evaluate `foo/two.dhall` it works correctly:

```
$ dhall --file foo/two.dhall
{ baz = "buz", foo = "bar" }
```

And when I try the same thing with `dhall-to-yaml --file` it fails to resolve the imports:
```
$ dhall-to-yaml --file foo/two.dhall

↳ ./one.dhall

Error: Missing file ./one.dhall

1│           ./one.dhall
2│ 
```
Copying the file to the directory I'm running these commands from results in `dhall-to-yaml` resolving the `./one.dhall` import.
```
$ cp foo/one.dhall .
$ dhall-to-yaml --file foo/two.dhall
foo: bar
baz: buz
```
I'd expect these to behave the same, but if there's a reason they shouldn't that's cool too.

### PR DESCRIPTION
**Pull request by sjakobi**
Fixes #1183.

### CODE REVIEW

File: dhall-json/src/Dhall/JSON.hs [hunk 1]
```diff
@@ -951,13 +953,18 @@ handleSpecialDoubles specialDoubleMode =
 codeToValue
   :: Conversion
   -> SpecialDoubleMode
-  -> Text  -- ^ Describe the input for the sake of error location.
+  -> Maybe FilePath  -- ^ The source file path. If no path is given, imports
+                     -- are resolved relative to the current directory.
```
Comment (sjakobi) [inline dhall-json/src/Dhall/JSON.hs]: @deepfire You introduced the `Text` input description in https://github.com/dhall-lang/dhall-json/pull/20. Is this change ok for you?

File: dhall/src/Dhall/Import.hs [hunk 2]
```diff
@@ -970,7 +971,13 @@ loadWith expr₀ = case expr₀ of
 
 -- | Resolve all imports within an expression
 load :: Expr Src Import -> IO (Expr Src X)
-load expression = State.evalStateT (loadWith expression) (emptyStatus ".")
+load = loadWithDir "."
+
+-- | Resolve all imports within an expression, importing relative to the given
+-- directory.
+loadWithDir :: FilePath -> Expr Src Import -> IO (Expr Src X)
```
Comment (sjakobi) [inline dhall/src/Dhall/Import.hs]: Would `loadWithRootDirectory` be a better name?

File: dhall-json/src/Dhall/JSON.hs [hunk 3]
```diff
@@ -951,13 +953,18 @@ handleSpecialDoubles specialDoubleMode =
...
   -> IO Value
-codeToValue conversion specialDoubleMode name code = do
-    parsedExpression <- Core.throws (Dhall.Parser.exprFromText (Data.Text.unpack name) code)
+codeToValue conversion specialDoubleMode mFilePath code = do
+    parsedExpression <- Core.throws (Dhall.Parser.exprFromText (fromMaybe "(stdin)" mFilePath) code)
 
-    resolvedExpression <- Dhall.Import.load parsedExpression
+    let rootDirectory = case mFilePath of
+            Nothing -> "."
+            Just fp -> System.FilePath.takeDirectory fp
```
Comment (sjakobi) [inline dhall-json/src/Dhall/JSON.hs]: Note that I didn't adopt `dhall`'s way of interpreting `--file -` as `stdin` here, as I think it's a bit redundant.

File: dhall/src/Dhall/Import.hs (hunk 2 above)
Comment (Gabriel439) [reply dhall/src/Dhall/Import.hs]: Maybe `loadRelativeTo`?

File: dhall-json/src/Dhall/JSON.hs (hunk 1 above)
Comment (Gabriel439) [reply dhall-json/src/Dhall/JSON.hs]: Yeah, this change is fine

File: dhall-json/src/Dhall/JSON.hs (hunk 3 above)
Comment (Gabriel439) [reply dhall-json/src/Dhall/JSON.hs]: Then we should probably change `dhall` to match, for consistency

**Comment by sjakobi** [comment]
AppVeyor fails due to

```
Uploading artifacts...
Maximum allowed artifact storage size of 50000 Mb will be exceeded.
```

I'd suggest clearing the cache, but this might be about the artifact deployment to GitHub?!

**Comment by Gabriel439** [comment]
@sjakobi: https://help.appveyor.com/discussions/problems/24655-maximum-allowed-artifact-storage-size-of-50000-mb-will-be-exceeded

**Pull request merged**